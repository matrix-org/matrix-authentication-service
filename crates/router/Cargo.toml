[package]
name = "mas-router"
version = "0.1.0"
authors = ["Quentin Gliech <quenting@element.io>"]
edition = "2021"
license = "Apache-2.0"

[dependencies]
axum = { version = "0.6.0-rc.5", default-features = false }
serde = { version = "1.0.147", features = ["derive"] }
serde_urlencoded = "0.7.1"
serde_with = "2.1.0"
url = "2.3.1"
ulid = "1.0.0"
