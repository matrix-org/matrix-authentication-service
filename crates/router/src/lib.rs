// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#![forbid(unsafe_code)]
#![deny(
    clippy::all,
    clippy::pedantic,
    clippy::str_to_string,
    rustdoc::broken_intra_doc_links
)]

pub(crate) mod endpoints;
pub(crate) mod traits;
mod url_builder;

pub use self::{endpoints::*, traits::Route, url_builder::UrlBuilder};

#[cfg(test)]
mod tests {
    use std::borrow::Cow;

    use url::Url;

    use super::*;

    #[test]
    fn test_relative_urls() {
        assert_eq!(
            OidcConfiguration.relative_url(),
            Cow::Borrowed("/.well-known/openid-configuration")
        );
        assert_eq!(Index.relative_url(), Cow::Borrowed("/"));
        assert_eq!(
            Login::and_continue_grant(42).relative_url(),
            Cow::Borrowed("/login?next=continue_authorization_grant&data=42")
        );
    }

    #[test]
    fn test_absolute_urls() {
        let base = Url::try_from("https://example.com/").unwrap();
        assert_eq!(Index.absolute_url(&base).as_str(), "https://example.com/");
        assert_eq!(
            OidcConfiguration.absolute_url(&base).as_str(),
            "https://example.com/.well-known/openid-configuration"
        );
    }
}
