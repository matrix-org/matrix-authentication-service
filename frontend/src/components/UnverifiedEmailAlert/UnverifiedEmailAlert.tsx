// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Alert } from "@vector-im/compound-web";
import { useEffect, useState } from "react";
import { useTranslation } from "react-i18next";

import { FragmentType, useFragment, graphql } from "../../gql";
import { Link } from "../Link";

import styles from "./UnverifiedEmailAlert.module.css";

export const UNVERIFIED_EMAILS_FRAGMENT = graphql(/* GraphQL */ `
  fragment UnverifiedEmailAlert_user on User {
    id
    unverifiedEmails: emails(first: 0, state: PENDING) {
      totalCount
    }
  }
`);

const UnverifiedEmailAlert: React.FC<{
  user?: FragmentType<typeof UNVERIFIED_EMAILS_FRAGMENT>;
}> = ({ user }) => {
  const data = useFragment(UNVERIFIED_EMAILS_FRAGMENT, user);
  const [dismiss, setDismiss] = useState(false);
  const { t } = useTranslation();

  const doDismiss = (): void => setDismiss(true);

  useEffect(() => {
    setDismiss(false);
  }, [data?.unverifiedEmails?.totalCount]);

  if (!data?.unverifiedEmails?.totalCount || dismiss) {
    return null;
  }

  return (
    <Alert
      type="critical"
      title={t("frontend.unverified_email_alert.title")}
      onClose={doDismiss}
      className={styles.alert}
    >
      {t("frontend.unverified_email_alert.text", {
        count: data.unverifiedEmails.totalCount,
      })}{" "}
      <Link to="/" hash="emails">
        {t("frontend.unverified_email_alert.button")}
      </Link>
    </Alert>
  );
};

export default UnverifiedEmailAlert;
