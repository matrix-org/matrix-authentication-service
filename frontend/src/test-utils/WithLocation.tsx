// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// @vitest-environment happy-dom

import { Provider } from "jotai";
import { useHydrateAtoms } from "jotai/utils";

import { appConfigAtom, locationAtom } from "../routing";

const HydrateLocation: React.FC<React.PropsWithChildren<{ path: string }>> = ({
  children,
  path,
}) => {
  useHydrateAtoms([
    [appConfigAtom, { root: "/" }],
    [locationAtom, { pathname: path }],
  ]);
  return <>{children}</>;
};

/**
 * Utility for testing components that rely on routing or location
 * For example any component that includes a <Link />
 * Eg:
 * ```
 * const component = create(
      <WithLocation path="/">
        <NavItem route={{ type: "profile" }}>Active</NavItem>
      </WithLocation>,
    );
 * ```
 */
export const WithLocation: React.FC<
  React.PropsWithChildren<{ path?: string }>
> = ({ children, path }) => {
  return (
    <Provider>
      <HydrateLocation path={path || "/"}>{children}</HydrateLocation>
    </Provider>
  );
};
