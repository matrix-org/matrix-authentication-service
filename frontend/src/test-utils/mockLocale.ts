// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { vi } from "vitest";

/**
 * Mock the locale on Intl.DateTimeFormat
 * To achieve stable formatted dates across environments
 * Defaults to `en-GB`
 */
export const mockLocale = (defaultLocale = "en-GB"): void => {
  const { DateTimeFormat } = Intl;
  vi.spyOn(Intl, "DateTimeFormat").mockImplementation(
    (
      locales?: Intl.LocalesArgument,
      options?: Intl.DateTimeFormatOptions | undefined,
    ) => new DateTimeFormat(locales || defaultLocale, options),
  );
};
