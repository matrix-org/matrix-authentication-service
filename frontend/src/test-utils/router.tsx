// Copyright 2024 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {
  RouterProvider,
  createMemoryHistory,
  createRootRoute,
  createRoute,
  createRouter,
} from "@tanstack/react-router";

const rootRoute = createRootRoute();
const index = createRoute({ getParentRoute: () => rootRoute, path: "/" });
rootRoute.addChildren([index]);

const router = createRouter({
  history: createMemoryHistory(),
  routeTree: rootRoute,
});

const routerReady = router.load();
// Because we also use this in the stories, we need to make sure we call this only in tests
if (import.meta.vitest) {
  // Make sure the router is ready before running any tests
  import("vitest").then(({ beforeAll }) =>
    beforeAll(async () => await routerReady),
  );
}

export const DummyRouter: React.FC<React.PropsWithChildren> = ({
  children,
}) => (
  /** @ts-expect-error: The router we pass doesn't match the "real" router, which is fine for tests */
  <RouterProvider router={router} defaultComponent={() => children} />
);
