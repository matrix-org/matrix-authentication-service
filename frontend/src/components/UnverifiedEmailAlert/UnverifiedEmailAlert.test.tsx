// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// @vitest-environment happy-dom

import { render, cleanup, fireEvent } from "@testing-library/react";
import { describe, it, expect, afterEach } from "vitest";

import { makeFragmentData } from "../../gql/fragment-masking";
import { DumbRouter } from "../../test-utils/router";

import UnverifiedEmailAlert, {
  UNVERIFIED_EMAILS_FRAGMENT,
} from "./UnverifiedEmailAlert";

describe("<UnverifiedEmailAlert />", () => {
  afterEach(cleanup);

  it("does not render a warning when there are no unverified emails", () => {
    const data = makeFragmentData(
      {
        id: "abc123",
        unverifiedEmails: {
          totalCount: 0,
        },
      },
      UNVERIFIED_EMAILS_FRAGMENT,
    );

    const { container } = render(
      <DumbRouter>
        <UnverifiedEmailAlert user={data} />
      </DumbRouter>,
    );

    expect(container).toMatchInlineSnapshot("<div />");
  });

  it("renders a warning when there are unverified emails", () => {
    const data = makeFragmentData(
      {
        id: "abc123",
        unverifiedEmails: {
          totalCount: 2,
        },
      },
      UNVERIFIED_EMAILS_FRAGMENT,
    );

    const { container } = render(
      <DumbRouter>
        <UnverifiedEmailAlert user={data} />
      </DumbRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it("hides warning after it has been dismissed", () => {
    const data = makeFragmentData(
      {
        id: "abc123",
        unverifiedEmails: {
          totalCount: 2,
        },
      },
      UNVERIFIED_EMAILS_FRAGMENT,
    );

    const { container, getByText, getByLabelText } = render(
      <DumbRouter>
        <UnverifiedEmailAlert user={data} />
      </DumbRouter>,
    );

    // warning is rendered
    expect(getByText("Unverified email")).toBeTruthy();

    fireEvent.click(getByLabelText("Close"));

    // no more warning
    expect(container).toMatchInlineSnapshot("<div />");
  });

  it("hides warning when count of unverified emails becomes 0", () => {
    const data = makeFragmentData(
      {
        id: "abc123",
        unverifiedEmails: {
          totalCount: 2,
        },
      },
      UNVERIFIED_EMAILS_FRAGMENT,
    );

    const { container, getByText, rerender } = render(
      <DumbRouter>
        <UnverifiedEmailAlert user={data} />
      </DumbRouter>,
    );

    // warning is rendered
    expect(getByText("Unverified email")).toBeTruthy();

    const newData = makeFragmentData(
      {
        id: "abc123",
        unverifiedEmails: {
          totalCount: 0,
        },
      },
      UNVERIFIED_EMAILS_FRAGMENT,
    );
    rerender(
      <DumbRouter>
        <UnverifiedEmailAlert user={newData} />
      </DumbRouter>,
    );

    // warning removed
    expect(container).toMatchInlineSnapshot("<div />");
  });

  it("shows a dismissed warning again when there are new unverified emails", () => {
    const data = makeFragmentData(
      {
        id: "abc123",
        unverifiedEmails: {
          totalCount: 2,
        },
      },
      UNVERIFIED_EMAILS_FRAGMENT,
    );

    const { container, getByText, getByLabelText, rerender } = render(
      <DumbRouter>
        <UnverifiedEmailAlert user={data} />
      </DumbRouter>,
    );

    // warning is rendered
    expect(getByText("Unverified email")).toBeTruthy();

    fireEvent.click(getByLabelText("Close"));

    // no more warning
    expect(container).toMatchInlineSnapshot("<div />");

    const newData = makeFragmentData(
      {
        id: "abc123",
        unverifiedEmails: {
          totalCount: 3,
        },
      },
      UNVERIFIED_EMAILS_FRAGMENT,
    );
    rerender(
      <DumbRouter>
        <UnverifiedEmailAlert user={newData} />
      </DumbRouter>,
    );

    // warning is rendered
    expect(getByText("Unverified email")).toBeTruthy();
  });
});
