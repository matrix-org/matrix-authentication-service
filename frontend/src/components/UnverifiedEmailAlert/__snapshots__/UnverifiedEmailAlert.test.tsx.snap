// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<UnverifiedEmailAlert /> > renders a warning when there are unverified emails 1`] = `
<div>
  <div
    class="_alert_1bz08_19 _alert_d86cd2"
    data-type="critical"
  >
    <svg
      aria-hidden="true"
      class="_icon_1bz08_57"
      fill="currentColor"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 17a.97.97 0 0 0 .713-.288A.968.968 0 0 0 13 16a.968.968 0 0 0-.287-.713A.968.968 0 0 0 12 15a.968.968 0 0 0-.713.287A.968.968 0 0 0 11 16c0 .283.096.52.287.712.192.192.43.288.713.288Zm0-4c.283 0 .52-.096.713-.287A.968.968 0 0 0 13 12V8a.967.967 0 0 0-.287-.713A.968.968 0 0 0 12 7a.968.968 0 0 0-.713.287A.967.967 0 0 0 11 8v4c0 .283.096.52.287.713.192.191.43.287.713.287Zm0 9a9.738 9.738 0 0 1-3.9-.788 10.099 10.099 0 0 1-3.175-2.137c-.9-.9-1.612-1.958-2.137-3.175A9.738 9.738 0 0 1 2 12a9.74 9.74 0 0 1 .788-3.9 10.099 10.099 0 0 1 2.137-3.175c.9-.9 1.958-1.612 3.175-2.137A9.738 9.738 0 0 1 12 2a9.74 9.74 0 0 1 3.9.788 10.098 10.098 0 0 1 3.175 2.137c.9.9 1.613 1.958 2.137 3.175A9.738 9.738 0 0 1 22 12a9.738 9.738 0 0 1-.788 3.9 10.098 10.098 0 0 1-2.137 3.175c-.9.9-1.958 1.613-3.175 2.137A9.738 9.738 0 0 1 12 22Z"
      />
    </svg>
    <div
      class="_content_1bz08_46"
    >
      <div
        class="_text-content_1bz08_53"
      >
        <p
          class="_typography_yh5dq_162 _font-body-md-semibold_yh5dq_64"
        >
          Unverified email
        </p>
        <p
          class="_typography_yh5dq_162 _font-body-sm-regular_yh5dq_40"
        >
          You have 2 unverified email addresses.
           
          <a
            aria-current="page"
            class="_link_1mzip_17 active"
            data-kind="primary"
            data-status="active"
            href="/#emails"
            rel="noreferrer noopener"
          >
            Review and verify
          </a>
        </p>
      </div>
    </div>
    <button
      aria-label="Close"
      class="_icon-button_16nk7_17"
      role="button"
      style="--cpd-icon-button-size: 32px;"
      tabindex="0"
    >
      <div
        class="_indicator-icon_133tf_26"
        style="--cpd-icon-button-size: 100%;"
      >
        <svg
          class="cpd-icon"
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.293 6.293a1 1 0 0 1 1.414 0L12 10.586l4.293-4.293a1 1 0 1 1 1.414 1.414L13.414 12l4.293 4.293a1 1 0 0 1-1.414 1.414L12 13.414l-4.293 4.293a1 1 0 0 1-1.414-1.414L10.586 12 6.293 7.707a1 1 0 0 1 0-1.414Z"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;
